### ISSUE TITLE
Performance issues with implicit returns in reduce functions

### ISSUE THREAD

**Issue by HendrikThePendric** [dhis2/notes#39]
Writing a reduce function like this looks really concise and clean:
```javascript
arr.reduce(
    (acc, { id, value }) => ({
        ...acc,
        [id]: value,
    }),
    {}
)
```

The alternative, looks a little bit less pretty:
```javascript
arr.reduce(
    (acc, { id, value }) => {
        acc[id]: value
        return acc
    },
    {}
)
```
However, the first version is making a shallow copy of the accumulator object on every iteration, and because of this I expected it to be a bit slower. 

I was interested in finding out how much slower this would actually be, so I made [a jsperf](https://jsperf.com/implicit-vs-explicit-return-in-reducer-fn/2) yesterday to find out. I was pretty shocked about the results: it's not just a little bit slower, it's a HUGE difference...

Because the differences were so extreme, I initially didn't trust the results. So I also created a [simple HTML page with a button](https://gistpreview.github.io/?1ced51f4632f22bbe25bc2979775079b) that tests the same thing, and confirms the results. And I also discussed things with @amcgee @Birkbjo and @Mohammer5, and they couldn't find anything wrong with my setup. 

@amcgee also prepared a [jsperf](https://jsperf.com/object-rest-spread/1) to compare object-rest-spread to parameter assignment in isolation.

Basically, everything points to the object spread operator and object assign being way slower than parameter assignment. So my advice would be: don't use this pattern in iterations. Especially in a reduce function there is no need to do this, if you just make sure that the `initialValue` of the accumulator object is "a fresh object", you really don't have to worry about doing mutations.

Let me know if you agree.

**Comment by amcgee** [comment]
Great work @HendrikThePendric!

It would be really interesting to analyze our code and see how often we’re using this slow pattern (it was *WAY* slower than I thought it would be in the perf tests) in iterable reduce functions.

In my [performance tests](https://jsperf.com/object-rest-spread/1) I found that a function performing explicit assignment performs comparably to direct assignment (1-10% overhead depending on the browser) and avoids the requirement to include a function body in the reduce handler:

```js
// fastest, somewhat clunky
list.reduce((out, item) => {
   out[item.key] = item.value
   return out
}, {})

// clean, second fastest (assuming setProperty is a common utility function)
const setProperty(obj, key, value) => {
  obj[key] = value
  return obj
}
list.reduce((out, item) => setProperty(out, item.key, item.value), {})

// cleanish, 10-20x slower!
list.reduce((out, item) => Object.assign(out, { [item.key]: item.value }), {})

// cleanish, 10-20x slower!
list.reduce((out, item) => ({
  ...out,
  [item.key]: item.value
}), {})
```

I agree 100% that this could be problematic, let's not do it!

**Comment by varl** [comment]
Good job taking the time to produce some numbers on common patterns! :100: It takes presence of mind to go from :thinking: to :straight_ruler:.

`Object.assign` and the Spread syntax produces a new object per iteration and copies over all the enumerable properties, so it is not a surprise that these are slower than _not_ creating a new object and _not_ copying over all properties on each iteration. The sheer _amount_ they are slower by is surprising, indeed.

The Spread syntax and `Object.assign` are very nice to use where immutability is required (e.g. reducers) but this is a very nice reminder that just because they make sense in one place, doesn't mean the pattern is applicable everywhere.

If you want to go even faster :tm:, remember that TurboFan optimizes by creating a type based on the object properties. So if you define `{ foo: 'bar' }`, V8 will create a class definition for that object which it will pass to TurboFan to use in loops. If that object gets a new property `{ foo: 'bar', fuz: 'buz' }` it fails the typecheck and gets passed back to Ignition. It in turn creates a new class def and passes it to TurboFan.

**Adding new properties to an object after instantiation breaks TurboFan optimizations.** This means that if it is possible to create a single object with all the properties defined (even if they are set to `undefined`), it will be faster to use that object. The more hints the engine gets, the better it can optimize.

https://jsperf.com/implicit-vs-explicit-return-in-reducer-fn/7

Firefox:
![2019-04-03-201042_977x500_scrot](https://user-images.githubusercontent.com/185449/55502311-980cf500-564c-11e9-85d6-01a4521815ec.png)

Chromium:
![2019-04-03-201122_978x500_scrot](https://user-images.githubusercontent.com/185449/55502396-c38fdf80-564c-11e9-9af9-44af218f5a17.png)

All in all this is a fine gotcha to write down and to have in mind! :+1: 

_N.B. I'm referring to the V8's **TurboFan** and **Ignition**, but Firefox has similar JIT optimization techniques in place._

**Comment by amcgee** [comment]
Wow, super deep dive nice @varl!

I was also surprised by the *amount* of slowdown when using object spread, but am thoroughly puzzled by the slowdown using Object.assign, since Object.assign **shouldn't** be creating new object clones each iteration (it just assigns to the first argument passed).  My guess is that the construction of an object just to pass as the second argument to Object.assign (`Object.assign(obj, { a: 'b' })`) is the bottleneck, not the cloning of all the properties.  Interesting stuff.

**Comment by varl** [comment]
I see that I should have been more precise and stated that some of the listed functions create _new_ Objects _and_ copies the properties. Some of them _only_ copies the properties. From what I can gather, all the slow ones **copy** and all the fast ones **assign**.

You are right; `Object.assign` does not create a new copy on each run, but it does get all enumerable properties and _copies_ them to the target.

Creating a new Object is quick (using `DOMHighResTimeStamp` I cannot even measure the speed), as is assigning a property to an Object, so I don't expect that to be the bottleneck, however copying properties from one object to another where the engines start getting into problems (which I can measure with `DOMHighResTimeStamp`), which both Spread and `Object.assign` does, under the hood.

`Object.assign` also runs any setters and getters, so on Objects which use those functions it can be very slow indeed. I don't _think_ native Spread calls those, but, most of the polyfills for Spread uses `Object.assign` so YMMV.

> `Object.assign(obj, { a: 'b' })`

So based on the above, this creates a new object for `{ a: 'b' }` (fast), and then it **copies** the prop `a` to `obj.a` (slow), invoking the setter for `a` in the process.

**Comment by amcgee** [comment]
Huh, the enumeration of the source object must be what's slow then.  Fascinating.

**Comment by HendrikThePendric** [comment]
@varl thanks for sharing that example that demonstrates optimisation techniques. Seems we all agree: **use assignment in reduce functions**. Now currently this is just a GitHub issue, but you mention it should be a gotcha. Shall I write a PR that adds a summary of this discussion to the gotchas section?

**Comment by HendrikThePendric** [comment]
Alternatively, we could add it as a decisions section. You could argue the case for both....

**Comment by janhenrikoverland** [comment]
I noticed the same thing some time ago and have added to and returned the same object in reduce iterations since then.

However, a guy called Sigurd Schneider from the v8 team had a presentation at last year's Web Rebels in Oslo (Edoardo, Jennifer, Viktor and I were there) and he was pretty clear that: "you write simple code, we fix the performance".

I still think we should go with the clunky one though. The perf difference is just too huge to ignore right now.

**Comment by varl** [comment]
>Now currently this is just a GitHub issue, but you mention it should be a gotcha. Shall I write a PR that adds a summary of this discussion to the gotchas section?

>Alternatively, we could add it as a decisions section. You could argue the case for both....

Absolutely. I consider this more of a gotcha than something we want to enforce through a decision in large part based on the quote @janhenrikoverland provided:

>"you write simple code, we fix the performance"

As a general rule I think it is heavy-handed to say never use object spread in a reduce function, because the problem is bound by the data set.

If we have 10'000 elements to reduce over and we run it on a field validation on every keystroke, it is important to maximise performance.

Yet if we have 10 elements, and we are able to cache the result? The more readable version might be the superiour choice, in that case.

So I think this is a "gotcha" we should have in mind, rather than a decision ("law") we can unilaterally apply to our entire code base in every situation.
